# Author: Jonathan Sprinkle
# Copyright (c) 2015 Arizona Board of Regents
# All rights reserved.
#
# Permission is hereby granted, without written agreement and without 
# license or royalty fees, to use, copy, modify, and distribute this
# software and its documentation for any purpose, provided that the 
# above copyright notice and the following two paragraphs appear in 
# all copies of this software.
#
# IN NO EVENT SHALL THE ARIZONA BOARD OF REGENTS BE LIABLE TO ANY PARTY 
# FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES 
# ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN 
# IF THE ARIZONA BOARD OF REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF 
# SUCH DAMAGE.
#
# THE ARIZONA BOARD OF REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, 
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY 
# AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER
# IS ON AN "AS IS" BASIS, AND THE ARIZONA BOARD OF REGENTS HAS NO OBLIGATION
# TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.

# Summary: 
# This YAML file includes the controller connectivity for ROS-based control
# through Gazebo. For more information and for the tutorials used to create
# this file, see
# http://gazebosim.org/tutorials/?tut=ros_control#Prerequisites

# controls the rear two tires based on individual motors
  # Publish all joint states -----------------------------------
  joint_state_controller:
    type: joint_state_controller/JointStateController
    publish_rate: 50  

  # Velocity Controllers ---------------------------------------

  # The left/right rear wheels are what moves the car. This
  # controller set uses PID for velocity set points
  joint1_velocity_controller:
    type: velocity_controllers/JointVelocityController
    joint: back_left_wheel_joint
    pid: {p: 100.0, i: 0.01, d: 10.0}

  joint2_velocity_controller:
    type: velocity_controllers/JointVelocityController
    joint: back_right_wheel_joint
    pid: {p: 100.0, i: 0.01, d: 10.0}

  # we are no longer using these controllers for angular rate
  # but we may reinstate them later
  #front_left_steering_controller:
  #  type: velocity_controllers/JointVelocityController
  #  joint: front_left_steering_joint
  #  pid: {p: 100.0, i: 0.01, d: 10.0}
  #front_right_steering_controller:
  #  type: velocity_controllers/JointVelocityController
  #  joint: front_right_steering_joint
  #  pid: {p: 100.0, i: 0.01, d: 10.0}

  # permits a set point for the steering (tire) angle, which 
  # is enforced by these position controllers. The value is
  # set based on the dynamics of the controller, not an
  # instantaneous set for the position
  front_left_steering_position_controller:
    type: position_controllers/JointPositionController
    joint: front_left_steering_joint
    pid: {p: 100.0, i: 0.01, d: 10.0}
  front_right_steering_position_controller:
    type: position_controllers/JointPositionController
    joint: front_right_steering_joint
    pid: {p: 100.0, i: 0.01, d: 10.0}

  #gazebo_ros_control:
  #  pid_gains:
  #    back_left_wheel_joint:
  #      p: 100.0
  #      i: 0.01
  #      d: 10.0
  #    back_right_wheel_joint:
  #      p: 100.0
  #      i: 0.01
  #      d: 10.0
  #    front_left_steering_joint:
  #      p: 100.0
  #      i: 0.01
  #      d: 10.0
  #    front_right_steering_joint:
  #      p: 100.0
  #      i: 0.01
  #      d: 10.0
